File: app/src/main/java/com/nimtego/itunes/presentation/base/BasePresenter.java
package com.nimtego.itunes.presentation.base;


import io.reactivex.annotations.NonNull;

public abstract class BasePresenter<V extends BaseContract.View,
        I extends BaseContract.Interactor>
        implements BaseContract.Presenter<V, I> {

    protected V view;
    protected I interactor;

    public BasePresenter(I interactor) {
        this.interactor = interactor;
    }

    @Override
    public void attach(@NonNull V view) {
        this.view = view;
    }

    @Override
    public void detach() {
        this.view = null;
        this.interactor.dispose();
    }

    @Override
    public V getView() {
        return view;
    }
}

File: app/src/main/java/com/nimtego/plectrum/presentation/mvp/presenters/base/BasePresenter.kt
package com.nimtego.plectrum.presentation.mvp.presenters.base

import android.util.Log
import com.arellomobile.mvp.MvpPresenter
import com.nimtego.plectrum.presentation.mvp.view.BaseView
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

abstract class BasePresenter<T : BaseView>(
        //private val disposableInteractor: DisposableInteractor
) : MvpPresenter<T>() {

    private val compositeDisposable = CompositeDisposable()

    protected val isViewAttached: Boolean get() = attachedViews.size > 0

    private val tag = javaClass.simpleName

    override fun attachView(view: T) {
        Log.i(this.tag, "attachView")
        super.attachView(view)
    }

    override fun detachView(view: T) {
        Log.i(this.tag, "detachView")
        super.detachView(view)
    }

    override fun onDestroy() {
        compositeDisposable.dispose()
    }

    protected fun Disposable.connect() {
        compositeDisposable.add(this)
    }

}